cmd/egress-v6-cni-plugin/cni.go

```diff
@@ -0,0 +1,608 @@
+		for _, r := range []netlink.Route{
+			{
+				LinkIndex: *containerVethIfIndex,
+				Dst: &net.IPNet{
+					IP:   net.IPv6zero,
+					Mask: net.CIDRMask(0, 128),
+				},
+				Scope: netlink.SCOPE_UNIVERSE,
+				Gw: netIPs[0],
+			},
+			//{
+			//	LinkIndex: *containerVethIfIndex,
+			//	Dst: &net.IPNet{
+			//		IP:   containerVethIfIPv6.IP.Mask(containerVethIfIPv6.Mask),
+			//		Mask: containerVethIfIPv6.Mask,
+			//	},
+			//	Scope: netlink.SCOPE_UNIVERSE,
+			//},
+		} {
+			// set up from container off-cluster IPv6 route (egress)
+			// all from container IPv6 traffic via host veth interface's link-local IPv6 address
+			if err := netlink.RouteAdd(&r); os.IsExist(err) {
+				// ignore this error
+			} else if err != nil {
+				return fmt.Errorf("failed to add route %v: %v", r, err)
+			}
+		}
+		return nil
+	})
+}
+
+func tearDownContainerIPv6Route(netns ns.NetNS, containerVethIfIPv6 *net.IPNet, containerVethIfName string) error {
+	containerVethIfIndex, err := getContainerVethIfIndex(netns, containerVethIfName)
+	if err != nil {
+		return err
+	}
+
+	netIPs, err := getHostVethIPv6ByContainerVethIfName(netns, containerVethIfName)
+	if err != nil {
+		return err
+	}
+
+	if len(netIPs) != 1 {
+		return fmt.Errorf("0 or more than 1 link local IPv6 addresses found in host veth interface")
+	}
+
+	return netns.Do(func(hostNS ns.NetNS) error {
+		for _, r := range []netlink.Route{
+			{
+				LinkIndex: *containerVethIfIndex,
+				Dst: &net.IPNet{
+					IP:   net.IPv6zero,
+					Mask: net.CIDRMask(0, 128),
+				},
+				Scope: netlink.SCOPE_UNIVERSE,
+				Gw: netIPs[0],
+			},
+			{
+				LinkIndex: *containerVethIfIndex,
+				Dst: &net.IPNet{
+					IP:   containerVethIfIPv6.IP.Mask(containerVethIfIPv6.Mask),
+					Mask: containerVethIfIPv6.Mask,
+				},
+				Scope: netlink.SCOPE_UNIVERSE,
+			},
+		} {
+			// delete from container off-cluster IPv6 route (egress)
+			// all from container IPv6 traffic via host veth interface's link-local IPv6 address
+			if err := netlink.RouteDel(&r); os.IsNotExist(err) {
+				// ignore this error
+			} else if err != nil {
+				return fmt.Errorf("failed to add route %v: %v", r, err)
+			}
+		}
+		return nil
+	})
+}
+
+func enableHostIPv6Forwarding() error {
+	var hostPrimaryIf string
+	err := ip.EnableIP6Forward()
+	if err != nil {
+		return err
+	}
+
+	// figure out host primary interface and set accept_ra = 2
+
+	primaryMAC, err := imds.GetMetaData("mac")
+	if err != nil {
+		return err
+	}
+
+	links, err := netlink.LinkList()
+	if err != nil {
+		return err
+	}
+
+	for _, link := range links {
+		if link.Attrs().HardwareAddr.String() == primaryMAC {
+			hostPrimaryIf = link.Attrs().Name
+			break
+		}
+	}
+
+	if len(hostPrimaryIf) > 0 {
+		var entry = "/proc/sys/net/ipv6/conf/" + hostPrimaryIf + "/accept_ra"
+
+		if content, err := os.ReadFile(entry); err == nil {
+			if bytes.Equal(bytes.TrimSpace(content), []byte("2")) {
+				return nil
+			}
+		}
+		return os.WriteFile(entry, []byte("2"), 0644)
```

Comment:
This can also be handled during container init: https://github.com/aws/amazon-vpc-cni-k8s/blob/master/cmd/aws-vpc-cni-init/main.go#L126